### PR TITLE
userbuffer: support fp8 buffer for individual overlap instance

### PR DESCRIPTION
**Pull request by erhoo82**
Make individual ReduceScatter overlap to use FP8 buffer so load the GEMM output in FP8 to save memory bandwidth cost.

### CODE REVIEW

File: transformer_engine/pytorch/module/base.py [hunk 1]
```diff
@@ -161,6 +159,7 @@ def add_ub(
         aggregate: int = 0,
         atomic_gemm: int = 0,
         is_reduce_scatter: int = 0,
+        fp8_buf: int = 0,
```
Comment (ksivaman) [inline transformer_engine/pytorch/module/base.py:162]:
```suggestion
        fp8_buf: bool = False,
```

File: transformer_engine/pytorch/module/base.py [hunk 2]
```diff
@@ -232,14 +231,17 @@ def add_ub(
...
+            num_sm = ub_cfg.get("num_sm", 16)
+            cga_size = ub_cfg.get("cga_size", 2)
+            num_splits = ub_cfg.get("num_splits", 4)
+            set_sm_margin = ub_cfg.get("set_sm_margin", 0)
+            aggregate = ub_cfg.get("aggregate", 0)
+            atomic_gemm = ub_cfg.get("atomic_gemm", 0)
             is_reduce_scatter = 1 if name in layers_reduce_scatter_overlap else 0
+            # Support FP8 userbuffer when (1) AllGather and (2) FP8-GEMM output ReduceScatter
+            fp8_buf = 1 if (name in layers_all_gather_overlap) or \
+                (ub_cfg.get("fp8_buf", 0) and name in methods["pipeline"]) else 0
```
Comment (ksivaman) [inline transformer_engine/pytorch/module/base.py:244]:
```suggestion
            fp8_buf = ((name in layers_all_gather_overlap) or
                      (ub_cfg.get("fp8_buf", False) and name in methods["pipeline"]))
```

**Review by ksivaman — CHANGES_REQUESTED** [review]
`bool`  is the better representation here, other than that, LGTM

**Comment by erhoo82** [comment]
Pushed the suggested changes.

**Review by ksivaman — APPROVED** [review]
Verified ubuf build locally

**Pull request merged**